### PR TITLE
feat(#191): Add another limiting strategy

### PR DESCRIPTION
**Pull request by vader-sama**
#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [X] documentation is changed or added
- [X] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by climba03003 — CHANGES_REQUESTED** [review]
It is actually forcing people to create more function in their custom store which should not be happen.

I think we can expose different official store for different strategy if they really need it.

File: test/global-rate-limit.test.js [hunk 1]
```diff
@@ -948,3 +948,77 @@ test('exposeHeadRoutes', async t => {
...
+      method: 'GET'
+    })
+
+    t.equal(first.statusCode, 200)
+
+    t.equal(second.statusCode, 429)
+    t.equal(second.headers['x-ratelimit-limit'], 1)
+    t.equal(second.headers['x-ratelimit-remaining'], 0)
+    t.equal(second.headers['x-ratelimit-reset'], 5)
+  } catch (error) {
```
Comment (mcollina) [inline test/global-rate-limit.test.js:981]: remove this try/catch

File: test/global-rate-limit.test.js [hunk 2]
```diff
@@ -948,3 +948,77 @@ test('exposeHeadRoutes', async t => {
...
```
Comment (mcollina) [inline test/global-rate-limit.test.js:1018]: remove the try/catch

File: test/route-rate-limit.test.js [hunk 3]
```diff
@@ -1322,3 +1322,89 @@ test('Allow custom timeWindow in preHandler', t => {
...
+      method: 'GET'
+    })
+
+    t.equal(first.statusCode, 200)
+
+    t.equal(second.statusCode, 429)
+    t.equal(second.headers['x-ratelimit-limit'], 1)
+    t.equal(second.headers['x-ratelimit-remaining'], 0)
+    t.equal(second.headers['x-ratelimit-reset'], 5)
+  } catch (error) {
```
Comment (mcollina) [inline test/route-rate-limit.test.js:1361]: remove the try/catch

File: test/route-rate-limit.test.js [hunk 4]
```diff
@@ -1322,3 +1322,89 @@ test('Allow custom timeWindow in preHandler', t => {
...
```
Comment (mcollina) [inline test/route-rate-limit.test.js:1404]: remove the try/catch

File: test/route-rate-limit.test.js [hunk 5]
```diff
@@ -1322,3 +1322,89 @@ test('Allow custom timeWindow in preHandler', t => {
...
+    t.error(error)
+  }
+
+  redis.flushall(noop)
+  redis.quit(noop)
```
Comment (mcollina) [inline test/route-rate-limit.test.js:1409]: move these into a `t.teardown` function

**Review by mcollina — COMMENTED** [review]
good work!

**Comment by vader-sama** [comment]
@climba03003 That's correct but they are coding a lot already (creating `incr` function), I think requested feature isn't in scope of this PR and It needs another PR; Don't you agree? 😉

**Comment by climba03003** [comment]
> @climba03003 That's correct but they are coding a lot already (creating `incr` function), I think requested feature isn't in scope of this PR and needs its own PR; Don't you agree? 😉

My suggest is preventing breaking change, it will break the people using `CustomStore` if they don't implement `incrAndRenew`. 


> @climba03003 That's correct but they are coding a lot already (creating `incr` function), I think requested feature isn't in scope of this PR and needs its own PR; Don't you agree? 😉

If we can provide a solution on `non-breaking-change`, I will go for that route. You can update the `storage` internal to provide the same feature.

I am against implementating one function for the store because it does not make sense on `CustomStore`.

**Comment by vader-sama** [comment]
> > @climba03003 That's correct but they are coding a lot already (creating `incr` function), I think requested feature isn't in scope of this PR and needs its own PR; Don't you agree? 😉
> 
> If we can provide a solution on `non-breaking-change`, I will go for that route. You can update the `storage` internal to provide the same feature.
> 
> I am against implementating extra function for the store because it does not make sense on `CustomStore` use-case.

But what's the solution then?

**Comment by climba03003** [comment]
> 

After https://github.com/fastify/fastify-rate-limit/blob/master/index.js#L78, you can pass the option inside the store.
`pluginComponent.store.continueExceeding = params.continueExceeding` 
Then, just update `.incr` function in the two store we provide.

**Comment by climba03003** [comment]
Even better solution is that, update the `constructor` of the two `store` and accept new param.
Then, update the `.child` method that pass the `param`
Finally, update the `.incr` function.
If also allow us to `enable` this feature per route.

**Comment by vader-sama** [comment]
@climba03003 What do you think now? (879ba4e57d2520)

**Comment by vader-sama** [comment]
@climba03003 What do you think now? (17dbd486b)

**Review by climba03003 — APPROVED** [review]
LGTM. Thanks.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by vader-sama** [comment]
@climba03003 Hi, this might be a weird question LOL; but is there any problem that my PR hasn't been merged yet?

**Review by Eomm — APPROVED** [review]

**Pull request merged**